src/Data/Id.hs

```diff
@@ -4,69 +4,172 @@
 #ifndef ghcjs_HOST_OS
-import           Database.PostgreSQL.Simple.FromField (FromField)
-import           Database.PostgreSQL.Simple.ToField (ToField)
+import           Database.PostgreSQL.Simple.FromField as PG (FromField)
+import           Database.PostgreSQL.Simple.ToField as PG (ToField)
+import           Data.Csv as Csv hiding(Name)
 #endif
-import           GHC.Generics (Generic)
+import           Flat as F
+import           GHC.TypeLits
 import           Test.QuickCheck
 import           Web.HttpApiData
 import           Web.PathPieces (PathPiece(..))
 
 newtype Id t = Id { unId :: UUID }
-  deriving
-  ( Eq, Ord, Generic, Read, Show, Data
+  deriving stock Data
+  deriving newtype
+  ( Eq, Ord, Binary
 #ifndef ghcjs_HOST_OS
-  , ToField, FromField
+  , PG.ToField, PG.FromField
 #endif
-  , PathPiece, FromJSON
-  , ToJSON, NFData, Hashable, FromJSONKey, ToJSONKey, ToSchema, ToParamSchema
-  , FromHttpApiData, ToHttpApiData )
+  , FromJSON, ToJSON, NFData, Hashable, FromJSONKey, ToJSONKey, ToSchema
+  , ToParamSchema, FromHttpApiData, ToHttpApiData )
 
 type role Id nominal
 
+makePrisms ''Id
+
+mkId :: forall s. UUID -> Id s
+mkId = coerce
+{-# INLINE mkId #-}
+
+instance KnownSymbol s => Show (Id s) where
+  show (Id v) = "Id-" <> symbolVal (Proxy @s) <> "-" <> show v
+
+instance Flat (Id s) where
+  encode = F.encode . UUID.toWords . coerce
+  decode = coerce . (\(a,b,c,d) -> UUID.fromWords a b c d) <$> F.decode
+  size = F.size . UUID.toWords . coerce
+
 instance Arbitrary (Id t) where
   arbitrary = fmap Id $ UUID.fromWords
     <$> arbitrary
     <*> arbitrary
     <*> arbitrary
     <*> arbitrary
 
-instance Binary (Id t) where
+instance PathPiece (Id t) where
+  fromPathPiece = coerce . UUID.fromText
+  toPathPiece = UUID.toText . coerce
 
-instance PathPiece UUID.UUID where
-  fromPathPiece = UUID.fromText
-  toPathPiece = UUID.toText
+#ifndef ghcjs_HOST_OS
 
-makePrisms ''Id
+instance Csv.FromField (Id tag) where
+  parseField = parseField
+    >=> fmap mkId . maybe (fail "invalid UUID in CSV") pure . UUID.fromText
+
+instance Csv.ToField (Id tag) where
+  toField = Csv.toField . UUID.toText . coerce
+
+#endif
 
 -- | This is a more \"explicit\" 'coerce' specifically for 'Id'.
 -- You are forced to explicitly specify the phantom types you are converting
 -- via the @TypeApplications@ compiler extension.
-coerceId :: forall a b. Id (Ambiguous a) -> Id (Ambiguous b)
+coerceId, unsafeIdTagConvert :: forall a b. Id (Ambiguous a) -> Id (Ambiguous b)
 coerceId = coerce
 {-# INLINE coerceId #-}
 
+-- just another name for 'coerceId'
+unsafeIdTagConvert = coerce
+{-# INLINE unsafeIdTagConvert #-}
+
 -- | Id corresponding to UUID nil. Useful for testing etc.
 nilId :: forall a.  Id (Ambiguous a)
 nilId = Id UUID.nil
 {-# INLINE nilId #-}
 -- https://kcsongor.github.io/ambiguous-tags/
 type family Ambiguous (a :: k) :: j where
   Ambiguous x = x
+
+----------------- IntId
+newtype IntId t = IntId { unIntId :: Integer }
```

Comment:
Do we actually need `Integer` in practice? Is it realistic to expect a situation where we `Int` is not enough?

Reply:
We have Integer Ids in Travolutionary. Moreover `int`'s from Postgres can be more than `Int`. 
I don't think that we can get much with this optimization...